src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java

```diff
@@ -222,60 +239,167 @@ private void endStream(RowStream<Row> stream, RoutingContext ctx,
         });
   }
 
-  Future<Tuple> ermLookup(RoutingContext ctx, String identifier) {
+  static Tuple parseErmTitle(JsonObject resource) {
+    String identifierValue = null;
+    JsonArray identifiers = resource.getJsonArray("identifiers");
+    for (int i = 0; i < identifiers.size(); i++) {
+      JsonObject identifier = identifiers.getJsonObject(i).getJsonObject("identifier");
+      identifierValue = identifier.getString("value");
+    }
+    UUID titleId = UUID.fromString(resource.getString("id"));
+    return Tuple.of(titleId, resource.getString("name"), identifierValue);
+  }
+
+  Future<Tuple> ermTitleLookup(RoutingContext ctx, String identifier) {
     // assuming identifier only has unreserved characters
+    // TODO .. this will match any type of identifier.
+    // what if there's more than one hit?
     String uri = "/erm/resource?match=identifiers.identifier.value&term=" + identifier;
-    Future<JsonArray> future = getRequest(webClient, ctx, uri)
+    return getRequest(ctx, uri)
+        .send()
+        .compose(res -> {
+          if (res.statusCode() != 200) {
+            return Future.failedFuture(uri + " returned " + res.statusCode());
+          }
+          return Future.succeededFuture(res.bodyAsJsonArray());
+        })
+        .map(ar -> ar.isEmpty() ? null : parseErmTitle(ar.getJsonObject(0)));
+  }
+
+  Future<Tuple> ermTitleLookup(RoutingContext ctx, UUID id) {
+    String uri = "/erm/resource/" + id;
+    return getRequest(ctx, uri)
+        .send()
+        .compose(res -> {
+          if (res.statusCode() != 200) {
+            return Future.failedFuture(uri + " returned " + res.statusCode());
+          }
+          return Future.succeededFuture(res.bodyAsJsonObject());
+        })
+        .map(EusageReportsApi::parseErmTitle);
+  }
+
+  Future<List<UUID>> ermPackageContentLookup(RoutingContext ctx, UUID id) {
+    // example: /erm/packages/dfb61870-1252-4ece-8f75-db02faf4ab82/content
+    String uri = "/erm/packages/" + id + "/content";
+    return getRequest(ctx, uri)
         .send()
         .compose(res -> {
           if (res.statusCode() != 200) {
             return Future.failedFuture(uri + " returned " + res.statusCode());
           }
           JsonArray ar = res.bodyAsJsonArray();
-          return Future.succeededFuture(ar);
+          List<UUID> list = new ArrayList<>();
+          for (int i = 0; i < ar.size(); i++) {
+            UUID kbTitleId = UUID.fromString(ar.getJsonObject(i).getJsonObject("pti")
+                .getJsonObject("titleInstance").getString("id"));
+            list.add(kbTitleId);
+          }
+          return Future.succeededFuture(list);
         });
```

Comment:
Please do not use a .compose without an async operation.

Reply:
No. Throws exception

Reply:
No, .map catches the exception.